src/sets.jl

```diff
@@ -240,3 +239,233 @@ function distance_to_set(d::DefaultDistance, v::AbstractVector{T}, s::MOI.Indica
+
+function projection_on_set(::DefaultDistance, ::MOI.EqualTo, z::Array{Float64}, dual=true)
+    return dual ? z : zeros(Float64, size(z))
+end
+
+function projection_on_set(::DefaultDistance, ::MOI.EqualTo, z::Float64, dual=true)
+    return dual ? [z] : zeros(Float64, 1)
+end
+
+"""
+    projection of vector `z` on Nonnegative cone i.e. K = R+
+"""
+function projection_on_set(::DefaultDistance, ::MOI.Nonnegatives, z::Array{Float64})
+    return max.(z, 0.0)
+end
+
+"""
+    projection of vector `z` on second order cone i.e. K = {(t, x) ∈ R+ × Rn |  ||x|| ≤ t }
+"""
+function projection_on_set(::DefaultDistance, ::MOI.SecondOrderCone, z::Array{Float64})
+    t = z[1]
+    x = z[2:length(z)]
+    norm_x = norm(x)
+    if norm_x <= t
+        return copy(z)
+    elseif norm_x <= -t
+        return zeros(Float64, size(z))
+    else
+        result = zeros(Float64, size(z))
+        result[1] = 1.0
+        result[2:length(z)] = x / norm_x
+        result *= (norm_x + t) / 2.0
+        return result
+    end
+end
+
+"""
+    projection of vector `z` on positive semidefinite cone i.e. K = S^n⨥
+"""
+function projection_on_set(::DefaultDistance, ::MOI.PositiveSemidefiniteConeTriangle, z::Array{Float64})
+    dim = Int64(floor(√(2*length(z))))
+    X = unvec_symm(z, dim)
+    λ, U = eigen(X)
+    return vec_symm(U * Diagonal(max.(λ,0)) * U')
+end
+
+"""
+    Returns a dim-by-dim symmetric matrix corresponding to `x`.
+
+    `x` is a vector of length dim*(dim + 1)/2, corresponding to a symmetric
+    matrix; the correspondence is as in SCS.
+    X = [ X11 X12 ... X1k
+          X21 X22 ... X2k
+          ...
+          Xk1 Xk2 ... Xkk ],
+    where
+    vec(X) = (X11, sqrt(2)*X21, ..., sqrt(2)*Xk1, X22, sqrt(2)*X32, ..., Xkk)
+"""
+function unvec_symm(x, dim)
+    X = zeros(dim, dim)
+    for i in 1:dim
+        for j in i:dim
+            X[j,i] = x[(i-1)*dim-Int(((i-1)*i)/2)+j]
+        end
+    end
+    X = X + X'
+    X /= √2
+    for i in 1:dim
+        X[i,i] /= √2
+    end
+    return X
+end
+
+"""
+    vec_symm(X)
+
+Returns a vectorized representation of a symmetric matrix `X`.
+Vectorization (including scaling) as per SCS.
+`vec(X) = (X11, sqrt(2)*X21, ..., sqrt(2)*Xk1, X22, sqrt(2)*X32, ..., Xkk)`
+"""
+function vec_symm(X)
+    X = copy(X)
+    X *= √2
+    for i in 1:size(X)[1]
+        X[i,i] /= √2
+    end
+    return X[tril(trues(size(X)))]
+end
+
+"""
+    projection_on_set(::DefaultDistance, cones::Array{<:MOI.AbstractSet}, z)
+
+Projection onto R^n x K^* x R_+
+ `cones` represents a convex cone K, and K^* is its dual cone
+"""
+function projection_on_set(::DefaultDistance, cones::Array{<:MOI.AbstractSet}, z)
+    @assert length(cones) == length(z)
+    return vcat([projection_on_set(cones[i], z[i]) for i in 1:length(cones)]...)
+end
+
+
+#  Derivative of the projection of vector `z` on MOI set `cone`
+#  projection_gradient_on_set[i,j] = ∂projection_on_set[i] / ∂z[j]   where `projection_on_set` denotes projection of `z` on `cone`
```

Comment:
these comments could be in the docstring of `projection_gradient_on_set`

Reply:
moved